File: bench/elixir/regex.ex
defmodule Bregex do
  import Bitwise
  # Generate N decimal strings; count full matches of \A[0-9]+\z. Checksum = count.
  def main do
    n = String.to_integer(System.get_env("BENCH_N") || "20000")
    re = ~r/\A[0-9]+\z/

    {_, count} =
      Enum.reduce(1..n, {123_456_789, 0}, fn _, {x, c} ->
        x2 = band(x * 1_103_515_245 + 12_345, 0x7FFFFFFF)
        s = Integer.to_string(x2)
        s = if rem(x2, 2) == 0, do: s <> "x", else: s
        {x2, c + if(Regex.match?(re, s), do: 1, else: 0)}
      end)

    IO.puts(count)
  end
end

File: bench/ruby/regex.rb
# Generate N decimal strings; count full matches of \A[0-9]+\z. Checksum = count.
n = (ENV["BENCH_N"] || "20000").to_i
re = /\A[0-9]+\z/
x = 123456789
count = 0
n.times do
  x = (x * 1103515245 + 12345) & 0x7FFFFFFF
  s = x.to_s
  s += "x" if x.even?
  count += 1 if s.match?(re)
end
puts count
